File: src-tauri/src/commands/decks.rs
use crate::db::decks::{self, DeckStats, DeckSummary, IndexCard};
use crate::db::entities::Entity;
use crate::db::DbState;
use crate::error::AppResult;
use tauri::State;

#[tauri::command]
pub fn create_deck(
    state: State<DbState>,
    space_id: String,
    title: String,
    exam_id: Option<String>,
) -> AppResult<Entity> {
    let conn = state.0.lock().unwrap();
    decks::create_deck(&conn, space_id, title, exam_id)
}

#[tauri::command]
pub fn list_decks(state: State<DbState>, space_id: String) -> AppResult<Vec<Entity>> {
    let conn = state.0.lock().unwrap();
    decks::list_decks(&conn, &space_id)
}

#[tauri::command]
pub fn list_deck_summaries(state: State<DbState>, space_id: String) -> AppResult<Vec<DeckSummary>> {
    let conn = state.0.lock().unwrap();
    decks::list_deck_summaries(&conn, &space_id)
}

#[tauri::command]
pub fn set_deck_exam(
    state: State<DbState>,
    deck_entity_id: String,
    exam_id: Option<String>,
) -> AppResult<()> {
    let conn = state.0.lock().unwrap();
    decks::set_deck_exam(&conn, &deck_entity_id, exam_id)
}

#[tauri::command]
pub fn deck_stats(state: State<DbState>, deck_entity_id: String) -> AppResult<DeckStats> {
    let conn = state.0.lock().unwrap();
    decks::deck_stats(&conn, &deck_entity_id)
}

#[tauri::command]
pub fn create_card(
    state: State<DbState>,
    deck_entity_id: String,
    front: String,
    back: String,
) -> AppResult<IndexCard> {
    let conn = state.0.lock().unwrap();
    decks::create_card(&conn, deck_entity_id, front, back)
}

#[tauri::command]
pub fn update_card(
    state: State<DbState>,
    card_id: String,
    front: Option<String>,
    back: Option<String>,
) -> AppResult<IndexCard> {
    let conn = state.0.lock().unwrap();
    decks::update_card(&conn, &card_id, front, back)
}

#[tauri::command]
pub fn delete_card(state: State<DbState>, card_id: String) -> AppResult<()> {
    let conn = state.0.lock().unwrap();
    decks::delete_card(&conn, &card_id)
}

#[tauri::command]
pub fn restore_card(state: State<DbState>, card_id: String) -> AppResult<()> {
    let conn = state.0.lock().unwrap();
    decks::restore_card(&conn, &card_id)
}

#[tauri::command]
pub fn list_cards(state: State<DbState>, deck_entity_id: String) -> AppResult<Vec<IndexCard>> {
    let conn = state.0.lock().unwrap();
    decks::list_cards(&conn, &deck_entity_id, false)
}

#[tauri::command]
pub fn study_queue(state: State<DbState>, deck_entity_id: String) -> AppResult<Vec<IndexCard>> {
    let conn = state.0.lock().unwrap();
    decks::study_queue(&conn, &deck_entity_id)
}

#[tauri::command]
pub fn review_card(state: State<DbState>, card_id: String, rating: String) -> AppResult<IndexCard> {
    let conn = state.0.lock().unwrap();
    decks::review_card(&conn, &card_id, decks::parse_rating(&rating)?)
}

#[tauri::command]
pub fn undo_review(state: State<DbState>, card_id: String) -> AppResult<IndexCard> {
    let conn = state.0.lock().unwrap();
    decks::undo_review(&conn, &card_id)
}

File: packages/frontend/src/lib/api/decks.ts
import { invoke } from "@tauri-apps/api/core";
import type { CardRating, DeckStats, DeckSummary, Entity, IndexCard } from "./types";

export function createDeck(spaceId: string, title: string, examId: string | null): Promise<Entity> {
  return invoke("create_deck", { spaceId, title, examId });
}

/// Every deck in the Space with its Exam and card counts, for the Decks page.
export function listDeckSummaries(spaceId: string): Promise<DeckSummary[]> {
  return invoke("list_deck_summaries", { spaceId });
}

/// Files the deck under an Exam, or under none with `null`.
export function setDeckExam(deckEntityId: string, examId: string | null): Promise<void> {
  return invoke("set_deck_exam", { deckEntityId, examId });
}

export function listDecks(spaceId: string): Promise<Entity[]> {
  return invoke("list_decks", { spaceId });
}

export function getDeckStats(deckEntityId: string): Promise<DeckStats> {
  return invoke("deck_stats", { deckEntityId });
}

export function createCard(deckEntityId: string, front: string, back: string): Promise<IndexCard> {
  return invoke("create_card", { deckEntityId, front, back });
}

export function updateCard(
  cardId: string,
  patch: { front?: string; back?: string },
): Promise<IndexCard> {
  return invoke("update_card", { cardId, ...patch });
}

export function deleteCard(cardId: string): Promise<void> {
  return invoke("delete_card", { cardId });
}

export function restoreCard(cardId: string): Promise<void> {
  return invoke("restore_card", { cardId });
}

export function listCards(deckEntityId: string): Promise<IndexCard[]> {
  return invoke("list_cards", { deckEntityId });
}

/// Cards to study now, in order: (re)learning, reviews, then new.
export function getStudyQueue(deckEntityId: string): Promise<IndexCard[]> {
  return invoke("study_queue", { deckEntityId });
}

export function reviewCard(cardId: string, rating: CardRating): Promise<IndexCard> {
  return invoke("review_card", { cardId, rating });
}

export function undoReview(cardId: string): Promise<IndexCard> {
  return invoke("undo_review", { cardId });
}
